### DISCUSSION TITLE
Different Image Signatures on multiple runs

### DISCUSSION THREAD

**Discussion by chkalch** [ImageMagick/ImageMagick#2970] [Help] [answered]
Hello,

two month ago I used MagickGetImageSignature to calculate checksums that are independent of metadata for all JPG-files on my file-server.

In the meantime I gathered JPG-files from other machines and compared their ImageSignatures to the ones from the previous run to find out which images already are on the server.

After no image of 28.308 could be found in 72.570 I assumed that there is a bug in my program but I came to the conclusion that MagickGetImageSignature returns different checksums for the same files.

Now I wonder if my understanding of this function, get a checksum that's like a fingerprint for the image itself, is correct. If the signature differs on subsequent runs its far away from being reliable. I used the same Ubuntu-System, but maybe it did some automatic updates in the meantime that affected MagickWand?

Can somebody please tell me how reliable this function is and if I can use it the way I need to?

**Comment by mbrijun** [comment, marked as answer]
It may help for you to compare the signatures generated by the same version of the program. There were some documented changes (I believe between version 6 and 7) in the algorithm to ensure the same signature is generates regardless of the bit-depth version of the program. In version 7 there were several bugs where the consistency of signatures would break, but the team were very quick to fix those.

From my experience, I would recommend choosing a version of IM and testing it on the same set of files with several bit-depth values Q8 vs. Q16 vs. Q16HDRI. If you get consistent results, stick with that version. In my case, I am using 7.0.9-26 Q8.

**Reply by chkalch** [reply]
Thank you for the reply.

Following your suggestion I tested different versions to calculate the signatures of an image.

The unknown version from Ubuntu repository that was installed when I did the first run:
f6136100e8c028d14c9b43584e7f9b02a47483c60b7a434c386039a0cb76ed0d

Current version from git, compiled with "--with-quantum-depth=8":
d14e28912b14fbd0683b3a4d2fed75ffb9c124b21fbdb40f0b2e90dfb0ad58aa

Current version from git, compiled with "--with-quantum-depth=16":
d14e28912b14fbd0683b3a4d2fed75ffb9c124b21fbdb40f0b2e90dfb0ad58aa

Current version from git, compiled with "--with-quantum-depth=32":
d14e28912b14fbd0683b3a4d2fed75ffb9c124b21fbdb40f0b2e90dfb0ad58aa

Latest version for Ubuntu focal in official repository (Package: libmagickwand-6.q16-dev, Version: 8:6.9.10.23+dfsg-2.1ubuntu11.1):
9dac86e8a759cab753251835a159989e84d2ddb3da6a2878743767bd2c3f6717

So its seems that Version 6 is the one in Ubuntu-repo and as my tests with different quantum depths on the current version on git results in three identical checksums I will go with this version. As a benefit I can compile this version even if something gets changed again, on the downside all signatures have to be recalculated.
